cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/Dag.java

```diff
@@ -0,0 +1,403 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.planner;
+
+import co.cask.cdap.etl.proto.Connection;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A DAG (directed acyclic graph).
+ */
+public class Dag {
+  private final Set<String> nodes;
+  private final Set<String> sources;
+  private final Set<String> sinks;
+  // stage -> outputs of that stage
+  private final SetMultimap<String, String> outgoingConnections;
+  // stage -> inputs for that stage
+  private final SetMultimap<String, String> incomingConnections;
+
+  public static Dag fromConnections(Collection<Connection> connections) {
+    SetMultimap<String, String> outgoingConnections = HashMultimap.create();
+    SetMultimap<String, String> incomingConnections = HashMultimap.create();
+    for (Connection connection : connections) {
+      outgoingConnections.put(connection.getFrom(), connection.getTo());
+      incomingConnections.put(connection.getTo(), connection.getFrom());
+    }
+    Dag dag = new Dag(outgoingConnections, incomingConnections);
+    dag.validate();
+    return dag;
+  }
+
+  public static Dag clone(Dag dag) {
+    return new Dag(dag.sources, dag.sinks, dag.outgoingConnections, dag.incomingConnections);
+  }
+
+  private Dag(SetMultimap<String, String> outgoingConnections,
+              SetMultimap<String, String> incomingConnections) {
+    this(calculateSources(outgoingConnections, incomingConnections),
+         calculateSinks(outgoingConnections, incomingConnections),
+         outgoingConnections, incomingConnections);
+  }
+
+  private Dag(Set<String> sources, Set<String> sinks,
+              SetMultimap<String, String> outgoingConnections,
+              SetMultimap<String, String> incomingConnections) {
+    Preconditions.checkArgument(!outgoingConnections.isEmpty(), "Cannot create a DAG without any connections");
+    Preconditions.checkArgument(!incomingConnections.isEmpty(), "Cannot create a DAG without any connections");
+    this.sources = new HashSet<>(sources);
+    this.sinks = new HashSet<>(sinks);
+    this.outgoingConnections = HashMultimap.create(outgoingConnections);
+    this.incomingConnections = HashMultimap.create(incomingConnections);
+    this.nodes = new HashSet<>();
+    nodes.addAll(outgoingConnections.keySet());
+    nodes.addAll(outgoingConnections.values());
```

Comment:
this is called by linearize() to make a copy of itself so that linearizing isn't destructive.  So we can't call it here otherwise we'll have an infinite loop.  This is why this particular constructor is private.